src/core/server/elasticsearch/client/log_query_and_deprecation.ts

```diff
@@ -0,0 +1,137 @@
+  }
+  return `[${error.name}]: ${error.message}`;
+}
+
+/**
+ * returns a string in format:
+ *
+ * status code
+ * method URL
+ * request body
+ *
+ * so it could be copy-pasted into the Dev console
+ */
+function getResponseMessage(event: DiagnosticResult): string {
+  const errorMeta = getRequestDebugMeta(event);
+  const body = errorMeta.body ? `\n${errorMeta.body}` : '';
+  return `${errorMeta.statusCode}\n${errorMeta.method} ${errorMeta.url}${body}`;
+}
+
+/**
+ * Returns stringified debug information from an Elasticsearch request event
+ * useful for logging in case of an unexpected failure.
+ */
+export function getRequestDebugMeta(event: DiagnosticResult): {
+  url: string;
+  body: string;
+  statusCode: number | null;
+  method: string;
+} {
+  const params = event.meta.request.params;
+  // definition is wrong, `params.querystring` can be either a string or an object
+  const querystring = convertQueryString(params.querystring);
+  return {
+    url: `${params.path}${querystring ? `?${querystring}` : ''}`,
+    body: params.body ? `${ensureString(params.body)}` : '',
+    method: params.method,
+    statusCode: event.statusCode!,
+  };
+}
+
+export const instrumentEsQueryAndDeprecationLogger = ({
+  logger,
+  client,
+  type,
+}: {
+  logger: Logger;
+  client: Client;
+  type: string;
+}) => {
+  const queryLogger = logger.get('query', type);
+  const deprecationLogger = logger.get('deprecation');
+  client.diagnostic.on('response', (error, event) => {
+    if (event) {
+      const opaqueId = event.meta.request.options.opaqueId;
+      const meta = opaqueId
+        ? {
+            http: { request: { id: event.meta.request.options.opaqueId } },
+          }
+        : undefined; // do not clutter logs if opaqueId is not present
+      let queryMsg = '';
+      if (error) {
+        if (error instanceof errors.ResponseError) {
+          queryMsg = `${getResponseMessage(event)} ${getErrorMessage(error)}`;
+        } else {
+          queryMsg = getErrorMessage(error);
+        }
+      } else {
+        queryMsg = getResponseMessage(event);
+      }
+
+      queryLogger.debug(queryMsg, meta);
+
+      if (event.warnings && event.warnings.length > 0) {
+        // Plugins can explicitly mark requests as originating from a user by
+        // removing the `'x-elastic-product-origin': 'kibana'` header that's
+        // added by default. User requests will be shown to users in the
+        // upgrade assistant UI as an action item that has to be addressed
+        // before they upgrade.
+        // Kibana requests will be hidden from the upgrade assistant UI and are
+        // only logged to help developers maintain their plugins
+        const requestOrigin =
+          (event.meta.request.params.headers != null &&
+            (event.meta.request.params.headers[
+              'x-elastic-product-origin'
+            ] as unknown as string)) === 'kibana'
```

Comment:
The ES team decided that each product should have it's own origin header like `{'x-elastic-product-origin': 'kibana'}`. ES does validation on this header for known Elastic products. They will include this header in the deprecation document created in the deprecations index. Then upgrade assistant will then filter out any deprecation documents with an 'x-elastic-product-origin' header an not show this to users because if it's our products then users can't fix it.

When a request comes from a user (e.g. when they use the console app) then we don't include this header so that users can see that they are using deprecated APIs. We haven't gotten as far yet... but once https://github.com/elastic/kibana/issues/120201 is done, we can ask teams to go through all their APIs, if there's an API that contains 100% user controlled content then they should mark that API call as coming from the user by not sending the x-elastic-product-origin header.

So from Kibana's perspective if we get a deprecation warning, it's either because us developers wrote a deprecated query or because users wrote it.